### PR TITLE
Split chat lines into pools and allow the player to filter them

### LINKED ISSUE #6354: Toggle notifications dialogs.
For the people who have their volumes down or need accessibility options, any way to toggle text notifications like a unit-ready or construction-complete or base-under-attack?
## <bountysource-plugin>

Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/4088016-toggle-notifications-dialogs?utm_campaign=plugin&utm_content=tracker%2F36085&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F36085&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### LINKED ISSUE #6400: Notify a player of gameplay events when all audio is muted
For example: When an enemy attacks your base, you get a EVA warning "Our base is under attack". Additionally, and optionally, this information could be offered with a basic text string somewhere on the screen (bottom left?), so if audio is muted for some reason the player still gets information that his/her base is under attack. Same thing with super weapons, airstrikes etc.

As a player, I'm forced to mute my computer audio in some situations (school, public spaces etc.). If I still would like to play OpenRA muted, I don't currently get information if my base is under attack or what is happening around the map. It's not very nice to see your base is half-destroyed just because you were not allowed to use audio in gameplay.
## <bountysource-plugin>

Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/4206715-notify-a-player-of-gameplay-events-when-all-audio-is-muted?utm_campaign=plugin&utm_content=tracker%2F36085&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F36085&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### LINKED ISSUE #16148: Removing debug msgs for q and w shortcuts
I personally use 'select unit by type' all the time as it is one of the best tools for unit micro management thus I get constant spam of msgs and chat noise
screenshot 2019-02-01 at 23 56 57
this makes it a lot harder harder to notice msgs written by spectators and other players as well as reading missed ones, I'd like it removed as it serves no significant purpose

### PR DESCRIPTION
**Pull request by dragunoff**
There's a lot going on here so I've tried to split it across commits in a logical way:
- Split chat lines into pools (e.g. System, Mission, Chat, Notifications, UI Feedback)
- Allow the player to toggle the display of certain chat pools (e.g. hide UI Feedback text)
- Add transcription support to speech notifications (e.g. display a chat line when EVA says "Unit lost")
- Add transcriptions to speech notifications in TD (add transcriptions to relevant EVA quotes)

Here are some screenshots to explain the idea:
ingame-chat
*In-game chat with filter options and transcriptions*

display-settings
*Display settings*

ingame-chat-common
*Filter options for mods without transcriptions*

Depends on #17663 (for the custom dropdown arrow image). The in-game chat filter panel is not necessary but I think it's a nice addition - bringing the filters in context and allowing the player to quickly change the settings. I had to add the cog glyph to D2k and TS - and ArtSrc PR will follow if this gets approved.

Closes #16148
Closes #6400
Closes #6354
Perhaps #3278 as well?
Provides the basis for solving #5775 and #12421

### CODE REVIEW

**Comment by dragunoff** [comment]
Rebased.

**Comment by netnazgul** [comment]
I'd suggest adding a short hint description of each option as it might not be entirely clear what is covered by which checkbox.

**Comment by abcdefg30** [comment]
Needs a rebase again. :/

**Comment by dragunoff** [comment]
Rebased. Confilct were only in the chrome image files (because of merging #17806).

**Comment by dragunoff** [comment]
Rebased after the merge of #17663

**Comment by dragunoff** [comment]
Update: rebased after the merge of #18366

**Comment by pchote** [comment]
What else is needed for this to fully solve #5775 and #12421?

I haven't had a chance to test this yet, but the images above suggest that spectators can already turn chat off?

**Comment by dragunoff** [comment]
@pchote The UI in the screenshots is a bit misleading perhaps. I included "Chat", "Mission" and "System" as disabled checkboxes so that it makes sense that the user is seeing stuff in the chat even when all available boxes are off. 

Filtering out spectator chat is made possible by this PR but there are some caveats with the UI. We need to disable the spec chat for spectators that have filtered it out - i.e. they should only have access to the "All" chat. This won't be a problem in replays.

If we want to allow filtering out the main chat as well, then we need to ensure that the user still has access to the chat window - for typing commands.

And there is an important distinction to be made between filtering chat messages and hiding the chat overlay.

File: OpenRA.Mods.Common/Widgets/Logic/Ingame/IngameChatLogic.cs
Comment (abcdefg30) [inline OpenRA.Mods.Common/Widgets/Logic/Ingame/IngameChatLogic.cs]: Style nit: Omit braces.

File: mods/cnc/audio/notifications.yaml
Comment (abcdefg30) [inline mods/cnc/audio/notifications.yaml:27]: `enmyunit` actually says "Enemy unit destroyed." Seems like you transcribed `enmyapp1`?

Comment (pchote) [reply mods/cnc/audio/notifications.yaml:27]: The announcement name is `EnemyUnitsApproaching` so this is just highlighting an earlier mistake?

**Review by abcdefg30 — DISMISSED** [review]
The concept lgtm ingame. One thing I wondered about were transcriptions and (more importantly) UI feedback selectors for spectators. I'm not sure if exposing that makes sense for spectators, but on the other hand I'm not sure if it is worth special casing something.

**Comment by dragunoff** [comment]
> 
> 
> The concept lgtm ingame. One thing I wondered about were transcriptions and (more importantly) UI feedback selectors for spectators. I'm not sure if exposing that makes sense for spectators, but on the other hand I'm not sure if it is worth special casing something.

Currently UI feedback includes only the messages coming from selection hotkeys ("Selected across screen / map") and these are relevant for spectators. They don't hear the announcer but I'd agree it's not worth special casing this. And fundamentally these are user settings that are also found in the settings panel and live in `settings.yaml` so it makes sense that they be the same regardless of player status.

**Comment by abcdefg30** [comment]
Ok, I'm fine with that, so just "fixup requested" for the older remarks.

File: OpenRA.Game/Game.cs
Comment (dragunoff) [inline OpenRA.Game/Game.cs:884]: One thing that bothers me with this implementation is that now there are a bunch of add-chat-line methods cluttering `Game.cs`. I thought about moving them to their own file but I couldn't figure out how to do it. IIRC it had something to do with passing the `OrderManager` instance. I didn't dig too deep into that though. And of course this can be done in a follow-up.

File: OpenRA.Mods.Common/Widgets/Logic/Ingame/IngameChatLogic.cs
Comment (dragunoff) [reply OpenRA.Mods.Common/Widgets/Logic/Ingame/IngameChatLogic.cs]: FIxed.

File: mods/cnc/audio/notifications.yaml
Comment (dragunoff) [reply mods/cnc/audio/notifications.yaml:27]: `EnemyUnitsApproaching` is not used anywhere and probably that's why it was not spotted until now. Should I take any action here?

**Comment by dragunoff** [comment]
Update: fixed style nit.

**Comment by dragunoff** [comment]
Update: Rebased and resolved conflicts.

**Comment by pchote** [comment]
I like the idea of the audio transcripts, but I'm sorry to say that I still find the UI presentation confusing and a bit of a deal breaker.

I understand the motivation for the ingame filter list and think this will be good UI once we actually have the ability to filter standard and team/spectator chat. But IMO right now it feels premature, and I expect the disabled options to cause a lot of confusion. I'm also not sure we should allow players to disable mission/system messages at all (and what is the difference between "System" and "UI feedback"?).

Likewise for the settings menu, where my confusion was amplified further by testing this PR in RA which doesn't have a "Transcriptions" option at all! Having to click a dropdown to access only one usable checkbox feels quite wrong.

Does anybody have a list of references for how other games handle this?

**Comment by pchote** [comment]
IMO it would be clearer if we could have a single "Show audio notifications in chat" checkbox in the audio settings, and extract the "Selected across screen" etc messages into notifications.yaml so they fall under the same setting. The ingame filtering could then be limited to all/team chat?

**Comment by dragunoff** [comment]
> 
> 
> I like the idea of the audio transcripts, but I'm sorry to say that I still find the UI presentation confusing and a bit of a deal breaker.
> 
> I understand the motivation for the ingame filter list and think this will be good UI once we actually have the ability to filter standard and team/spectator chat. But IMO right now it feels premature, and I expect the disabled options to cause a lot of confusion. I'm also not sure we should allow players to disable mission/system messages at all (and what is the difference between "System" and "UI feedback"?).
> 
> Likewise for the settings menu, where my confusion was amplified further by testing this PR in RA which doesn't have a "Transcriptions" option at all! Having to click a dropdown to access only one usable checkbox feels quite wrong.
> 

Mission and system messages can't be disabled. 

"UI Feedback" currently is used only for _"Selected across map/screen"_ messages - i.e. UI events that don't give any other visual or audible feedback. The "System" pool is for events such as game start/end, player connects/disconnets, etc. and for debug.

I added the disabled checkboxes for completeness (each checkbox corresponds to a chat pool) and to make it clear to the player why they are seeing stuff in chat even when they've unchecked all boxes. 

Initially I added a "Spectators" pools for toggling spectator chat but decided to leave it off for now as its a bit trickier with the UI for that feature.

> 
> 
> IMO it would be clearer if we could have a single "Show audio notifications in chat" checkbox in the audio settings, and extract the "Selected across screen" etc messages into notifications.yaml so they fall under the same setting. The ingame filtering could then be limited to all/team chat?

_"Selected across..."_ messages are not audible so they are not the same thing (well, there is a beep sound but no voice line). Lumping them together with the narrator's lines would make it impossible to see these notifications without seeing also all of the narrator's lines. Which may be OK... Perhaps other should chime in here and we should have a look at how other games do it.

**Comment by pchote** [comment]
In any case this isn't critical for the next release, so dropping from the next release as part of a wider cleanup to try and unstall progress there.

**Review by Mailaender — APPROVED** [review]
Works as promised and code looks good.

**Comment by dragunoff** [comment]
I still very much have interest in finishing this. I'd be fine with removing the controversial UI and leaving that for a follow-up when we can also implement a spectator chat filter.

**Comment by dragunoff** [comment]
@pchote How about something like this for the UI:
image

**Comment by pchote** [comment]
How about putting them all on the same line?

Screenshot 2020-11-21 at 22 50 01

This would avoid breaking the y spacing between each row.

FWIW I still think it would make sense for this setting to live on the Audio tab instead.

**Comment by dragunoff** [comment]
@pchote I don't mind putting them on line line, will do. To me it makes more sense to put it in the "Display" tab because it is really about things displaying on the screen. Or perhaps it's time to seriously consider overhauling the settings panel to support arbitrary amount of settings and tabs (not in this PR of course). I'd be interested to work on that for next+1.

**Comment by dragunoff** [comment]
Closing in favor of #18905

**Pull request closed without merging**
